src/www/app/electron_main.ts

```diff
@@ -32,69 +29,62 @@ import {getLocalizationFunction, main} from './main';
 import {NativeNetworking} from './net';
 import {AbstractUpdater} from './updater';
 import {UrlInterceptor} from './url_interceptor';
+import {throwIfIpcError} from './util';
 import {VpnInstaller} from './vpn_installer';
 
-const isWindows = os.platform() === 'win32';
-const isLinux = os.platform() === 'linux';
+const isWindows = window.os.platform === 'win32';
+const isLinux = window.os.platform === 'linux';
 const isOsSupported = isWindows || isLinux;
 
 const interceptor = new UrlInterceptor();
-ipcRenderer.on('add-server', (e: Event, url: string) => {
-  interceptor.executeListeners(url);
-});
+window.ipc.onAddServer(interceptor.executeListeners);
 
-ipcRenderer.on('localizationRequest', (e: Event, localizationKeys: string[]) => {
+window.ipc.onLocalizationRequest(localizationKeys => {
   const localize = getLocalizationFunction();
   if (!localize) {
     console.error('Localization function not available.');
-    ipcRenderer.send('localizationResponse', null);
+    window.ipc.sendLocalizationResponse(null);
     return;
   }
   const localizationResult: {[key: string]: string} = {};
   for (const key of localizationKeys) {
     localizationResult[key] = localize(key);
   }
-  ipcRenderer.send('localizationResponse', localizationResult);
+  window.ipc.sendLocalizationResponse(localizationResult);
 });
 
 // Pushes a clipboard event whenever the app window receives focus.
 class ElectronClipboard extends AbstractClipboard {
   constructor() {
     super();
-    ipcRenderer.on('push-clipboard', this.emitEvent.bind(this));
+    window.ipc.onPushClipboard(this.emitEvent.bind(this));
   }
 
   getContents() {
-    return Promise.resolve(clipboard.readText());
+    return navigator.clipboard.readText();
   }
 }
 
 class ElectronUpdater extends AbstractUpdater {
   constructor() {
     super();
-    ipcRenderer.on('update-downloaded', this.emitEvent.bind(this));
+    window.ipc.onUpdateDownloaded(this.emitEvent.bind(this));
   }
 }
 
 class ElectronVpnInstaller implements VpnInstaller {
   public async installVpn(): Promise<void> {
-    const err = await ipcRenderer.invoke('install-outline-services');
-
-    // catch custom errors (even simple as numbers) does not work for ipcRenderer:
-    // https://github.com/electron/electron/issues/24427
-    if (err !== ErrorCode.NO_ERROR) {
-      throw new OutlinePluginError(err);
-    }
+    const err = await window.ipc.sendInstallOutlineServices();
+    throwIfIpcError(err);
```

Comment:
Perhaps we should move the throw to the preload script? Does that work?

Reply:
I tried but unfortunately it's not working, because they are from two different webpack entries, the classes are not shareable, compilation will fail. We can only use a very limited set of classes in preload.

Reply:
Do we have to share the error classes? Why not create a new `IpcError` in the preload script, or something like that

Reply:
All error type information and properties will be cleared out when you throw that in preload because of context isolation: https://github.com/electron/electron/issues/25596#issuecomment-700281408

Reply:
This is the sanitized error information (you can see there is nothing in it no matter what error you are throwing out, `throw 123;`, `throw new Error('test')` or `throw new OutlinePluginError(...)`): 
<img width="808" alt="image" src="https://user-images.githubusercontent.com/93548144/182450566-b5ac4396-2c2d-4575-a380-45a360cfc053.png">